redbox-core/redbox/graph/chat.py

```diff
@@ -0,0 +1,157 @@
+
+
+@chain
+def set_chat_method(state: ChainState):
+    """
+    Choose an approach to chatting based on the current state
+    """
+    log.debug("Selecting chat method")
+    number_of_docs = len(state["documents"])
+    if number_of_docs == 0:
+        selected_tool = ChatRoute.chat
+    elif number_of_docs == 1:
+        selected_tool = ChatRoute.chat_with_docs
+    else:
+        selected_tool = ChatRoute.chat_with_docs_map_reduce
+    log.info(f"Selected: {selected_tool} for execution")
+    return {"route_name": selected_tool}
+
+
+def build_llm_map_chain(
+    llm: BaseChatModel, tokeniser: Encoding, env: Settings, system_prompt: str, question_prompt: str
+) -> Runnable:
+    return (
+        make_chat_prompt_from_messages_runnable(
+            system_prompt=system_prompt,
+            question_prompt=question_prompt,
+            input_token_budget=env.ai.context_window_size - env.llm_max_tokens,
+            tokeniser=tokeniser,
+        )
+        | llm
+        | StrOutputParser()
+        | RunnableLambda(lambda s: {"intermediate_docs": [Document(page_content=s)]})
+    )
+
+
+def get_chat_with_docs_graph(
+    llm: BaseChatModel,
+    all_chunks_retriever: VectorStoreRetriever,
+    tokeniser: Encoding,
+    env: Settings,
+    debug: bool = False,
+) -> CompiledGraph:
+    app = StateGraph(ChainState)
+
+    app.add_node("get_chat_docs", build_get_docs(env, all_chunks_retriever))
+    app.add_node("set_chat_prompt_args", set_prompt_args)
+    app.add_node("set_chat_method", set_chat_method)
+
+    app.add_node("no_docs_available", get_no_docs_available(env))
+    app.add_node(
+        "llm",
+        build_llm_chain(
+            llm, tokeniser, env, env.ai.chat_with_docs_system_prompt, env.ai.chat_with_docs_question_prompt
+        ),
+    )
+    app.add_node(ChatRoute.chat_with_docs_map_reduce, get_chat_with_docs_map_reduce_graph(llm, tokeniser, env, debug))
+
+    app.add_edge(START, "get_chat_docs")
+    app.add_edge("get_chat_docs", "set_chat_prompt_args")
+    app.add_edge("set_chat_prompt_args", "set_chat_method")
+    app.add_conditional_edges(
+        "set_chat_method",
+        lambda state: state["route_name"],
+        {
+            ChatRoute.chat: "no_docs_available",
+            ChatRoute.chat_with_docs: "llm",
+            ChatRoute.chat_with_docs_map_reduce: ChatRoute.chat_with_docs_map_reduce,
+        },
+    )
+    app.add_edge(ChatRoute.chat_with_docs_map_reduce, "set_chat_prompt_args")
+    return app.compile(debug=debug)
+
+
+def get_chat_with_docs_map_reduce_graph(
+    llm: BaseChatModel, tokeniser: Encoding, env: Settings, debug: bool = False
+) -> CompiledGraph:
+    app = StateGraph(ChatMapReduceState)
+
+    app.add_node(
+        "llm_map", build_llm_map_chain(llm, tokeniser, env, env.ai.map_system_prompt, env.ai.chat_map_question_prompt)
+    )
+    app.add_node(
+        "reduce",
+        build_reduce_docs_step(
+            TokenTextSplitter(
+                model_name="gpt-4",
```

Comment:
Should this be `gpt-4o` instead?

Reply:
yeah, I'm not sure, this should be an env var so we switch it as needed